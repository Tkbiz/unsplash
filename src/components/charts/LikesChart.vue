<script>
import {Line} from 'vue-chartjs'

export default {
  extends: Line,
  props: {
    photoLikes: {
      type: Object,
      required: true
    }
  },
  mounted() {
    this.renderChart({
      labels: this.photoLikes.historical.values.map(date => date.date),
      datasets: [
        {
          label: 'Photo\'s likes',
          data: this.photoLikes.historical.values.map(value => value.value),
          fill: false,
          borderColor: '#57a773',
          backgroundColor: '#57a773',
          borderWidth: 1
        }
      ]
    }, {
      scales: {
        yAxes: [{
          ticks: {
            beginAtZero: true
          },
          gridLines: {
            display: true
          }
        }],
        xAxes: [{
          gridLines: {
            display: false
          }
        }]
      },
      legend: {
        display: true
      },
      responsive: true,
      maintainAspectRatio: false
    })
  }
}
</script>
