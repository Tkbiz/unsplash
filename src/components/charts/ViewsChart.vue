<script>
import {Bar} from 'vue-chartjs'

export default {
  extends: Bar,
  props: {
    photoViews: {
      type: Object,
      required: true
    }
  },
  mounted() {
    this.renderChart({
      labels: this.photoViews.historical.values.map(date => date.date),
      datasets: [
        {
          label: 'Photo\'s views',
          data: this.photoViews.historical.values.map(value => value.value),
          fill: false,
          borderColor: '#57a773',
          backgroundColor: '#57a773',
          borderWidth: 1
        }
      ]
    }, {
      scales: {
        yAxes: [{
          ticks: {
            beginAtZero: true
          },
          gridLines: {
            display: true
          }
        }],
        xAxes: [{
          gridLines: {
            display: false
          }
        }]
      },
      legend: {
        display: true
      },
      responsive: true,
      maintainAspectRatio: false
    })
  }
}
</script>
