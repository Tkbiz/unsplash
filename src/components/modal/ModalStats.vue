<template>
  <v-col cols="12"
         class="py-0 stats-block">
    <v-row>
      <h3 class="stats-title">
        Photo's statistic:
      </h3>
      <div v-if="photoStats.downloads"
           class="col-12 stats-info">
        Download:
        {{ photoStats.downloads.total }}
      </div>
      <div v-if="photoStats.likes"
           class="col-12 stats-info">
        Likes:
        {{ photoStats.likes.total }}
      </div>
      <div v-if="photoStats.views"
           class="col-12 stats-info">
        View:
        {{ photoStats.views.total }}
      </div>
      <div v-if="tags"
           class="col-12 stats-info">
        Tags:
        <span v-for="tag in tags"
              :key="tag.title">
                {{ tag.title }}
              </span>
      </div>
    </v-row>
  </v-col>
</template>

<script>
export default {
  name: 'ModalStats',
  props: {
    photoStats: {
      type: Object,
      required: true
    },
    tags: {
      type: Array,
      required: true
    }
  }
}
</script>
