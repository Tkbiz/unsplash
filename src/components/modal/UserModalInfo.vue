<template>
  <v-col cols="12"
         class="py-0 mb-9">
    <v-row
        v-if="userCard"
        align="center">
      <v-avatar
          v-if="userCard.profile_image"
          @click.prevent="openUserProfile(userCard.id)"
          size="84">
        <img
            :src="userCard.profile_image.large ? userCard.profile_image.large : '@/assets/images/no-image.png'"
            alt="John">
      </v-avatar>
      <div>
        <div v-if="userCard.name"
             @click.prevent="openUserProfile(userCard.id)"
             class="avatar-title">
          {{ userCard.name }}
        </div>
        <div v-if="userCard.location"
             class="avatar-nick">
          {{ userCard.location }}
        </div>
      </div>
    </v-row>
  </v-col>
</template>

<script>

export default {
  name: 'UserModalInfo',
  props: {
    userCard: {
      type: Object,
      required: true
    }
  },
  methods: {
    async openUserProfile(id) {
      if (this.userCard.username) {
        await this.$store.dispatch('fetchUserInfo', this.userCard.username)
        await this.$router.push(`/user-account/${id}`)
      }
    }
  }
}
</script>
