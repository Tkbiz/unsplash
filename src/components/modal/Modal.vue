<template>
  <v-layout
      row
      justify-center>
    <v-dialog
        content-class="custom-dialog"
        persistent
        v-model="showModal">
      <v-btn
          class="btn-close"
          icon
          @click.prevent="closeModal">
        <v-icon>
          mdi-close
        </v-icon>
      </v-btn>
      <div class="dialog-body">
        <v-col class="image-block">
          <v-img
              v-if="card.urls"
              class="main-img"
              :src="card.urls.regular ? card.urls.regular : '@/assets/images/no-image.png'">
          </v-img>
          <v-col class="action-block">
            <v-btn
                class="mr-2 btn-like"
                :class="{'btn-active': isLiked}"
                icon
                @click.prevent="addLike(card.id)">
              <v-icon>mdi-heart</v-icon>
            </v-btn>
            <v-btn
                class="btn-add"
                :class="{'btn-active': isAddedToCollection}"
                icon
                @click.prevent="openCollectionsList">
              <v-icon>mdi-plus</v-icon>
            </v-btn>
            <v-alert
                class="custom-alert"
                v-if="showMessage"
                type="warning">
              Please log in to do it!
            </v-alert>
            <v-alert
                class="collections-alert"
                v-model="showCollectionsList"
                text>
              <v-card>
                <v-card-title
                    v-for="collection in userCollections"
                    :key="collection.id"
                    @click.prevent="addToCollection(collection.id, card.id)"
                    class="headline cursor-pointer">
                  {{ collection.title }}
                </v-card-title>
              </v-card>
            </v-alert>
          </v-col>
        </v-col>
        <v-col v-if="card.description"
               cols="12"
               class="py-0 stats-block">
          <v-row>
            <div class="col-12 stats-info stats-description">
              {{ card.description }}
            </div>
          </v-row>
        </v-col>
        <UserModalInfo
            :userCard="card.user"/>
        <ModalStats
            :photoStats="photoStats"
            :tags="card.tags"/>
        <v-col cols="12 mb-6">
          <v-row>
            <div v-if="photoStats.views">
              <ViewsChart
                  :photoViews="photoStats.views"/>
            </div>
            <div v-if="photoStats.likes">
              <LikesChart
                  :photoLikes="photoStats.likes"/>
            </div>
          </v-row>
        </v-col>
        <v-col v-if="samePhotos"
               cols="12"
               class="pa-0">
          <v-row justify="start">
            <div class="block-img"
                 v-for="item in samePhotos"
                 :key="item.id">
              <img
                  class="image"
                  :src="item.urls.thumb ? item.urls.thumb : '@/assets/images/no-image.png'"
                  alt="card image">
            </div>
          </v-row>
        </v-col>
        <Loader v-if="loading"/>
        <v-col v-if="sameNewPhotos"
               cols="12"
               class="py-0">
          <v-row
              justify="center">
            <v-btn
                icon
                @click.prevent="showMoreImages">
              <v-icon
                  class="dots">
                mdi-dots-horizontal
              </v-icon>
            </v-btn>
          </v-row>
        </v-col>
      </div>
    </v-dialog>
  </v-layout>
</template>

<script>
import ViewsChart from '@/components/charts/ViewsChart'
import LikesChart from '@/components/charts/LikesChart'
import ModalStats from '@/components/modal/ModalStats'
import UserModalInfo from '@/components/modal/UserModalInfo'

export default {
  name: 'Modal',
  props: {
    visibleModal: {
      type: Boolean,
      required: true
    },
    card: {
      type: Object,
      required: true
    },
    samePhotos: {
      type: Array,
      required: true
    },
    tagName: {
      type: String,
      required: true
    },
    photoStats: {
      type: Object,
      required: true
    },
    userName: {
      type: String
    }
  },
  components: {
    UserModalInfo,
    ModalStats,
    ViewsChart,
    LikesChart
  },
  data: () => ({
    page: 1,
    sameNewPhotos: [],
    loading: false,
    showMessage: false,
    showCollectionsList: false
  }),
  methods: {
    closeModal() {
      this.showModal = false
    },
    async showMoreImages() {
      this.loading = true
      this.page++
      this.sameNewPhotos = await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page
      })
      this.sameNewPhotos.map(newPhoto => this.samePhotos.push(newPhoto))
      this.loading = false
    },
    async addLike(id) {
      if (this.isLoggedIn) {
        if (this.isLiked !== true) {
          await this.$store.dispatch('addLike', id)
        }
      } else {
        this.showMessage = true
        setTimeout(() => {
          this.showMessage = false
        }, 2000)
      }
    },
    async openCollectionsList() {
      if (this.isLoggedIn) {
        this.showCollectionsList = true
        await this.$store.dispatch('fetchUserCollections', this.userName)
      } else {
        this.showMessage = true
        setTimeout(() => {
          this.showMessage = false
        }, 2000)
      }
    },
    async addToCollection(collectionId, photoId) {
      await this.$store.dispatch('addToCollection', {collectionId, photoId})
      this.showCollectionsList = false
    }
  },
  computed: {
    showModal: {
      get() {
        return this.visibleModal
      },
      set(value) {
        if (!value) {
          this.$emit('close')
        }
      }
    },
    isLoggedIn() {
      return this.$store.getters.accessToken
    },
    userCollections() {
      return this.$store.getters.userCollections
    },
    isAddedToCollection() {
      return this.$store.getters.isAddedToCollection
    },
    isLiked() {
      return this.$store.getters.isLiked
    }
  }
}
</script>
