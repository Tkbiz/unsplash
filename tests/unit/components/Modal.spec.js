import {mount, createLocalVue} from '@vue/test-utils'
import Vue from 'vue'
import Vuetify from 'vuetify'
import Modal from '@/components/modal/Modal'
import ViewsChart from '@/components/charts/ViewsChart'
import LikesChart from '@/components/charts/LikesChart'
import ModalStats from '@/components/modal/ModalStats'
import UserModalInfo from '@/components/modal/UserModalInfo'
import Loader from '@/components/app/Loader'
import Vuex from 'vuex'

Vue.use(Vuetify)
Vue.component('Loader', Loader)

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Modal.vue', () => {
  let vuetify
  let wrapper
  let actions
  let getters
  let store
  const spy = {}

  beforeEach(() => {
    vuetify = new Vuetify()

    getters = {
      isLoggedIn: () => false,
      userCollections: () => ([]),
      isAddedToCollection: () => false,
      isLiked: () => false
    }
    actions = {
      searchPhotos: jest.fn(),
      addLike: jest.fn(),
      addToCollection: jest.fn(),
      fetchUserCollections: jest.fn()
    }
    store = new Vuex.Store({
      getters,
      actions
    })

    const el = document.createElement('div')
    el.setAttribute('data-app', true)
    document.body.appendChild(el)

    spy.closeModal = jest.spyOn(Modal.methods, 'closeModal')

    wrapper = mount(Modal, {
      localVue,
      vuetify,
      propsData: {
        visibleModal: true,
        card: {
          user: {},
          tags: []
        },
        samePhotos: [],
        tagName: 'tag name',
        photoStats: {},
        userName: 'user name'
      },
      data: () => ({
        page: 1,
        sameNewPhotos: [],
        loading: false,
        showMessage: false,
        showCollectionsList: false
      }),
      store
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  it('renders ViewsChart', () => {
    expect(wrapper.exists(ViewsChart)).toBe(true)
  })

  it('renders LikesChart', () => {
    expect(wrapper.exists(LikesChart)).toBe(true)
  })

  it('renders ModalStats', () => {
    expect(wrapper.exists(ModalStats)).toBe(true)
  })

  it('renders UserModalInfo', () => {
    expect(wrapper.exists(UserModalInfo)).toBe(true)
  })

  it('renders Loader', () => {
    expect(wrapper.exists(Loader)).toBe(true)
  })

  it('renders props when passed', () => {
    const visibleModal = true
    const card = {
      user: {},
      tags: []
    }
    const samePhotos = []
    const tagName = 'tag name'
    const photoStats = {}
    const userName = 'user name'
    expect(wrapper.props()).toEqual({
      visibleModal,
      card,
      samePhotos,
      tagName,
      photoStats,
      userName
    })
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders isLoggedIn using computed options', () => {
    expect(wrapper.vm.isLoggedIn).toBe(undefined)
  })

  it('renders userCollections using computed options', () => {
    expect(JSON.stringify(wrapper.vm.userCollections)).toBe(JSON.stringify([]))
  })

  it('renders isAddedToCollection using computed options', () => {
    expect(wrapper.vm.isAddedToCollection).toBe(false)
  })

  it('renders isLiked using computed options', () => {
    expect(wrapper.vm.isLiked).toBe(false)
  })

  it('renders showModal using computed options', () => {
    expect(wrapper.vm.showModal).toBe(true)
  })

  it('when button is clicked modal will be closed', () => {
    wrapper.find('.btn-close').trigger('click')
    wrapper.vm.$nextTick()

    expect(spy.closeModal).toHaveBeenCalled()
  })

  it("dispatches a searchPhotos action when button is clicked", async () => {
    const tagName = 'tag name'
    const page = 2

    wrapper.find('.dialog-body .py-0 .v-btn').trigger('click')
    await wrapper.vm.$nextTick()

    expect(actions.searchPhotos.mock.calls).toHaveLength(1)
    expect(actions.searchPhotos.mock.calls[0][1]).toEqual({tagName, page})
    expect(actions.searchPhotos).toHaveBeenCalled()
  })

  it("dispatches a addLike action when button is clicked", async () => {
    const id = 'id'
    wrapper.find('.action-block .btn-like').trigger('click')
    await wrapper.vm.$nextTick()

    if (wrapper.vm.isLoggedIn) {
      if (wrapper.vm.isLiked !== true) {
        expect(actions.addLike.mock.calls).toHaveLength(1)
        expect(actions.addLike.mock.calls[0][1]).toEqual({id})
        expect(actions.addLike).toHaveBeenCalled()
      }
    }
  })

  it("dispatches a fetchUserCollections action when button is clicked", async () => {
    const userName = 'user name'
    wrapper.find('.action-block .btn-add').trigger('click')
    await wrapper.vm.$nextTick()

    if (wrapper.vm.isLoggedIn) {
      wrapper.setData({showCollectionsList: true})

      expect(actions.fetchUserCollections.mock.calls).toHaveLength(1)
      expect(actions.fetchUserCollections.mock.calls[0][1]).toEqual({userName})
      expect(actions.fetchUserCollections).toHaveBeenCalled()
    }
  })

  it("dispatches a addToCollection action when button is clicked", async () => {
    const collectionId = 'collectionId'
    const photoId = 'photoId'
    wrapper.find('.action-block .btn-add').trigger('click')
    await wrapper.vm.$nextTick()

    if (wrapper.vm.showCollectionsList) {
      expect(actions.addToCollection.mock.calls).toHaveLength(1)
      expect(actions.addToCollection.mock.calls[0][1]).toEqual({collectionId, photoId})
      expect(actions.addToCollection).toHaveBeenCalled()
    }
  })
})
