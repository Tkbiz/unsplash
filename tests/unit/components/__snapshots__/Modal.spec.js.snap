// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.vue should match snapshot 1`] = `
<div class="layout row justify-center">
  <div role="dialog" class="v-dialog__container"></div>
</div>
`;
